src/c.rs

```diff
@@ -0,0 +1,33 @@
+// Copyright 2016-2019 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+//! C types.
+//!
+//! The libc crate provide the C types for most, but not all, targets that
+//! *ring* supports.
+
+use libc;
+
+pub(crate) type size_t = libc::size_t;
```

Comment:
We could just use `usize` instead throughout *ring* as suggested in https://github.com/briansmith/ring/pull/837#issuecomment-501877466. Note that many of the places *ring* uses `size_t` it should really be using `uintptr_t`; however, [BoringSSL also makes this "mistake" in thinking that `sizeof(size_t) == sizeof(void *)`](https://github.com/briansmith/ring/blob/8fa0cf5b0169b1ca99ef4b0787fbf5f02ce387f2/crypto/fipsmodule/modes/gcm.c#L57), so it might just be easier to assume `size_t == uintptr_t` in which case (as `libc::uintptr_t == usize`), we can just use `usize` throughout *ring*.

This could be addressed now, or in a future PR. We might also want to wait until https://github.com/rust-lang/libc/issues/1400 is resolved.

Reply:
> Note that many of the places _ring_ uses `size_t` it should really be using `uintptr_t`

Could you provide an example?

Rust has made a mistake where it is only well defined on platforms where `uintptr_t` and `size_t` and `usize` are the same type; whenever `size_t` and `uintptr_t` are incompatible then Rust can't work given how `usize` is defined. I'm hoping a future edition of Rust will correct this. This is actually the reason I prefer to keep using `size_t` everywhere in Rust where `size_t` is used in the C code.

> [BoringSSL also makes this "mistake" in thinking that sizeof(size_t) == sizeof(void *)](https://github.com/briansmith/ring/blob/8fa0cf5b0169b1ca99ef4b0787fbf5f02ce387f2/crypto/fipsmodule/modes/gcm.c#L57)

I think you may be misunderstanding the example you pointed out; the test in that code fragment has to do with whether 64-bit arithmetic is native or not, nothing about pointers. https://boringssl.googlesource.com/boringssl/+/8f574c37dae935a29bff56e750101e5b354503de/crypto/fipsmodule/bn/exponentiation.c#895 would be an example where `size_t` is used instead of `uintptr_t` in BoringSSL. I believe we've eliminated all such confusion in *ring* already; see #68, #151, and in particular #101.